src/datomish/datom.cljc

```diff
@@ -0,0 +1,136 @@
+;; This Source Code Form is subject to the terms of the Mozilla Public
+;; License, v. 2.0. If a copy of the MPL was not distributed with this
+;; file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+;; Purloined from DataScript.
+
+(ns datomish.datom)
+
+(declare hash-datom equiv-datom seq-datom val-at-datom nth-datom assoc-datom)
+
+(deftype Datom [e a v tx added]
+  #?@(:cljs
+      [IHash
+       (-hash [d] (or (.-__hash d)
+                      (set! (.-__hash d) (hash-datom d))))
+       IEquiv
+       (-equiv [d o] (and (instance? Datom o) (equiv-datom d o)))
+
+       ISeqable
+       (-seq [d] (seq-datom d))
+
+       ILookup
+       (-lookup [d k] (val-at-datom d k nil))
+       (-lookup [d k nf] (val-at-datom d k nf))
+
+       IIndexed
+       (-nth [this i] (nth-datom this i))
+       (-nth [this i not-found] (nth-datom this i not-found))
+
+       IAssociative
+       (-assoc [d k v] (assoc-datom d k v))
+
+       IPrintWithWriter
+       (-pr-writer [d writer opts]
+                   (pr-sequential-writer writer pr-writer
+                                         "#datascript/Datom [" " " "]"
+                                         opts [(.-e d) (.-a d) (.-v d) (.-tx d) (.-added d)]))
+       ]
+      :clj
+      [Object
+       (hashCode [d] (hash-datom d))
+
+       clojure.lang.IHashEq
+       (hasheq [d] (hash-datom d))
+
+       clojure.lang.Seqable
+       (seq [d] (seq-datom d))
+
+       clojure.lang.IPersistentCollection
+       (equiv [d o] (and (instance? Datom o) (equiv-datom d o)))
+       (empty [d] (throw (UnsupportedOperationException. "empty is not supported on Datom")))
+       (count [d] 5)
+       (cons [d [k v]] (assoc-datom d k v))
+
+       clojure.lang.Indexed
+       (nth [this i]           (nth-datom this i))
+       (nth [this i not-found] (nth-datom this i not-found))
+
+       clojure.lang.ILookup
+       (valAt [d k] (val-at-datom d k nil))
+       (valAt [d k nf] (val-at-datom d k nf))
+
+       clojure.lang.Associative
+       (entryAt [d k] (some->> (val-at-datom d k nil) (clojure.lang.MapEntry k)))
+       (containsKey [e k] (#{:e :a :v :tx :added} k))
+       (assoc [d k v] (assoc-datom d k v))
+       ]))
+
+(defn ^Datom datom
+  ([e a v tx]       (Datom. e a v tx true))
+  ([e a v tx added] (Datom. e a v tx added)))
+
+(defn datom? [x] (instance? Datom x))
+
+(defn- hash-datom [^Datom d]
```

Comment:
Copy-pasta from DataScript.  I think this will be because in the DS store, Datoms compare equal without regard to `tx` (and `added` is implicitly `true`).  I'll just cull pieces of this that aren't used.